logging: panic: runtime error: invalid memory address or nil pointer dereference

**Client**

e.g. Cloud Logging

**Environment**

e.g. Alpine Docker on GKE

**Go Environment**

$ go version
$ go env

**Code**

https://chromium-review.googlesource.com/c/infra/infra/+/4769930/2/go/src/infra/go.mod

**Expected behavior**

Don't panic.

**Actual behavior**

https://logs.chromium.org/logs/chromium/buildbucket/cr-buildbucket/8773108843220167633/+/u/Build_all_without_remote_cache/stdout

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xab79b8]

goroutine 177 [running]:
cloud.google.com/go/logging.(*loggerRetryer).Retry(0xc03d12d130, {0x1020d20, 0xc09477e630})
	cloud.google.com/go/logging@v1.8.0/logging.go:275 +0x78
github.com/googleapis/gax-go/v2.invoke({0x102b460, 0xc0992178f0}, 0xc0404e5e90, {0xc03d12d140, {0x0, 0x0, 0x0}, {0x0, 0x0}, 0xdf8475800}, ...)
	github.com/googleapis/gax-go/v2@v2.12.0/invoke.go:108 +0x21d
github.com/googleapis/gax-go/v2.Invoke({0x102b460, 0xc0992178f0}, 0xc0404e5e70?, {0xc0847b7900, 0x3, 0xc099217830?})
	github.com/googleapis/gax-go/v2@v2.12.0/invoke.go:50 +0xdd
cloud.google.com/go/logging/apiv2.(*gRPCClient).WriteLogEntries(0xc000050540, {0x102b428?, 0xc099217830?}, 0xc055371420, {0xc03d12d160, 0x1, 0xc0404e5f10?})
	cloud.google.com/go/logging@v1.8.0/apiv2/logging_client.go:360 +0x192
cloud.google.com/go/logging/apiv2.(*Client).WriteLogEntries(...)
	cloud.google.com/go/logging@v1.8.0/apiv2/logging_client.go:225
cloud.google.com/go/logging.(*Logger).writeLogEntries(0xc000720c00, {0xc03221b800, 0x3e8, 0x500})
	cloud.google.com/go/logging@v1.8.0/logging.go:738 +0x23b
cloud.google.com/go/logging.(*Client).Logger.func2({0xd15ec0?, 0xc07f8d21f8?})
	cloud.google.com/go/logging@v1.8.0/logging.go:298 +0x45
google.golang.org/api/support/bundler.(*Bundler).handle(0xc0000ea420, 0x0?)
	google.golang.org/api@v0.136.0/support/bundler/bundler.go:324 +0x4a
created by google.golang.org/api/support/bundler.(*Bundler).enqueueCurBundle
	google.golang.org/api@v0.136.0/support/bundler/bundler.go:179 +0x14d
```

**Additional context**

Started after upgrading to v1.8.0

--- a/logging/logging.go
+++ b/logging/logging.go
@@ -91,6 +91,9 @@ const (
 	// timeout is to allow clients to degrade gracefully if underlying logging
 	// service is temporarily impaired for some reason.
 	defaultWriteTimeout = 10 * time.Minute
+
+	// Part of the error message when the payload contains invalid UTF-8 characters.
+	utfErrorString = "string field contains invalid UTF-8"
 )
 
 var (
@@ -264,12 +267,28 @@ type loggerRetryer struct {
 	defaultRetryer gax.Retryer
 }
 
+func newLoggerRetryer() gax.Retryer {
+	// Copied from CallOptions.WriteLogEntries in apiv2/logging_client.go.
+	d := gax.OnCodes([]codes.Code{
+		codes.DeadlineExceeded,
+		codes.Internal,
+		codes.Unavailable,
+	}, gax.Backoff{
+		Initial:    100 * time.Millisecond,
+		Max:        60000 * time.Millisecond,
+		Multiplier: 1.30,
+	})
+
+	r := &loggerRetryer{defaultRetryer: d}
+	return r
+}
+
 func (r *loggerRetryer) Retry(err error) (pause time.Duration, shouldRetry bool) {
 	s, ok := status.FromError(err)
 	if !ok {
 		return r.defaultRetryer.Retry(err)
 	}
-	if s.Code() == codes.Internal && strings.Contains(s.Message(), "string field contains invalid UTF-8") {
+	if s.Code() == codes.Internal && strings.Contains(s.Message(), utfErrorString) {
 		return 0, false
 	}
 	return r.defaultRetryer.Retry(err)
@@ -734,8 +753,7 @@ func (l *Logger) writeLogEntries(entries []*logpb.LogEntry) {
 	ctx, cancel := context.WithTimeout(ctx, defaultWriteTimeout)
 	defer cancel()
 
-	r := &loggerRetryer{}
-	_, err := l.client.client.WriteLogEntries(ctx, req, gax.WithRetry(func() gax.Retryer { return r }))
+	_, err := l.client.client.WriteLogEntries(ctx, req, gax.WithRetry(newLoggerRetryer))
 	if err != nil {
 		l.client.error(err)
 	}

--- a/logging/logging.go
+++ b/logging/logging.go
@@ -288,7 +288,7 @@ func (r *loggerRetryer) Retry(err error) (pause time.Duration, shouldRetry bool)
 	if !ok {
 		return r.defaultRetryer.Retry(err)
 	}
-	if s.Code() == codes.Internal && strings.Contains(s.Message(), utfErrorString) {
+	if strings.Contains(s.Message(), utfErrorString) {
 		return 0, false
 	}
 	return r.defaultRetryer.Retry(err)

--- a/logging/logging_unexported_test.go
+++ b/logging/logging_unexported_test.go
@@ -18,6 +18,7 @@ package logging
 
 import (
 	"encoding/json"
+	"fmt"
 	"net/http"
 	"net/url"
 	"testing"
@@ -31,7 +32,42 @@ import (
 	"google.golang.org/api/support/bundler"
 	mrpb "google.golang.org/genproto/googleapis/api/monitoredres"
 	logtypepb "google.golang.org/genproto/googleapis/logging/type"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 )
+
+func TestLoggerRetryer_Retry(t *testing.T) {
+	for _, tst := range []struct {
+		name      string
+		err       error
+		wantRetry bool
+	}{
+		{
+			name:      "non_status_no_retry",
+			err:       fmt.Errorf("non-API error, do not retry"),
+			wantRetry: false,
+		},
+		{
+			name:      "invalid_utf_no_retry",
+			err:       status.Error(codes.Internal, utfErrorString),
+			wantRetry: false,
+		},
+		{
+			// Just testing one of the configured codes to ensure the default
+			// retryer is triggered.
+			name:      "unavailable_retry",
+			err:       status.Error(codes.Unavailable, "Unavailable"),
+			wantRetry: true,
+		},
+	} {
+		t.Run(tst.name, func(t *testing.T) {
+			_, gotRetry := newLoggerRetryer().Retry(tst.err)
+			if gotRetry != tst.wantRetry {
+				t.Errorf("Retry(%v) = shouldRetry got %v want %v", tst.err, gotRetry, tst.wantRetry)
+			}
+		})
+	}
+}
 
 func TestLoggerCreation(t *testing.T) {
 	const logID = "testing"

